### PR TITLE
Fix ssh connection error

### PR DESCRIPTION
**Pull request by SparkSnail**
SSH client has a max number of open channels for a connection, if we set the number of trialCurrency too big, our ssh client will exec command using ssh frequently, then we will meet the error of `Error: (SSH) Channel open failure: open failed`. 
Refactor the code, set one connection has a max trial concurrency, when the number of trial reach the ssh connection restriction, will create a new ssh connection to exec trial commands.
Refer:
https://github.com/mscdex/ssh2/issues/219

### CODE REVIEW

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 1]
```diff
@@ -58,8 +58,12 @@ import { mkDirP } from '../../common/utils';
  */
 @component.Singleton
 class RemoteMachineTrainingService implements TrainingService {
-    private machineSSHClientMap: Map<RemoteMachineMeta, Client>;
+    private machineSSHClientMap: Map<RemoteMachineMeta, Client>; //set cluster ssh client
+    private trialSSHClientMap: Map<string, Client>; //set trial ssh client
     private trialJobsMap: Map<string, RemoteMachineTrialJobDetail>;
+    private connectionNumberMap: Map<Client, number>; // set connection number for every ssh client
+    private connectionMetaMap: Map<Client, RemoteMachineMeta>; //set connection meta for every ssh client
```
Comment (yds05) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: Suggest to remove 'set' word in line 61-65

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 2]
```diff
@@ -58,8 +58,12 @@ import { mkDirP } from '../../common/utils';
...
+    private maxSSHTrialNumber: number; // every ssh client has a max connection number
```
Comment (yds05) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: readonly? If yes, suggest to rename the var to UPPER case, like MAX_TRIAL_NUMBER_PER_SSHCONNECTION

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts (hunk 1 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: refactor data structure, don't need to use these two map.

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts (hunk 2 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: updated to MAX_TRIAL_NUMBER_PER_SSHCONNECTION

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 3]
```diff
@@ -115,6 +118,92 @@ class RemoteMachineTrainingService implements TrainingService {
...
+        }
+
+        deferred.resolve();
+        return deferred.promise;
+    }
+
+    /**
+     * Create a new ssh connection client by trial job id
+     */
+    public async createSSHClient(trial: RemoteMachineTrialJobDetail): Promise<void> {
```
Comment (yds05) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: The naming of this function is confused.

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 4]
```diff
@@ -115,6 +118,92 @@ class RemoteMachineTrainingService implements TrainingService {
...
+        }).connect(connectConfig);
+      
+        return deferred.promise;
+    }
+    
+    /**
+     * If a trial is finished, retrieve the connection resource
+     * @param trial 
+     */
+    public retrieveSSHClient(trial: RemoteMachineTrialJobDetail): void {
```
Comment (yds05) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: The naming of this function is confused.

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts (hunk 3 above)
Comment (yds05) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: You should also consider put SSH client initialization into SSH connection manager.

Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: 1.changed `createSSHClient` to `scheduleTrialSSHClient`  

2.move create new ssh client logic to SSHClientManager

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts (hunk 4 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: changed to `releaseTrialSSHClient`

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 5]
```diff
@@ -115,6 +118,50 @@ class RemoteMachineTrainingService implements TrainingService {
         }
         this.log.info('Remote machine training service exit.');
     }
+    
+    /**
+     * give trial a ssh connection
+     * @param trial 
+     */
+    public async scheduleTrialSSHClient(trial: RemoteMachineTrialJobDetail): Promise<void> {
```
Comment (yds05) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: Suggest to rename the function to 'allocateSSHClientForTrial'

Comment (yds05) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: You didn't use await in your function. So no need to mark the function as async and return Promise

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 6]
```diff
@@ -115,6 +118,50 @@ class RemoteMachineTrainingService implements TrainingService {
...
+        }
+        let sshClientManager: SSHClientManager | undefined = this.machineSSHClientMap.get(trial.rmMeta);
+        if(!sshClientManager) {
+            throw new Error(`remoteSSHClient not initialized`);
+        }
+        let sshClient: Client | undefined = sshClientManager.getAvailableSSHClient();
+        if(!sshClient) {
+            //create a new connection
+            sshClient = await sshClientManager.initNewSSHClient();
+        }
```
Comment (yds05) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: I think the caller of sshClientManager doesnt need to know when to init. Just put the logic of iniNewSSHClient() into getAvailableSSHClient

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 7]
```diff
@@ -115,6 +118,50 @@ class RemoteMachineTrainingService implements TrainingService {
...
+        let sshClientManager: SSHClientManager | undefined = this.machineSSHClientMap.get(trial.rmMeta);
+        if(!sshClientManager) {
+            throw new Error(`sshClientManager not initialized`);
+        }
+        const conn: Client | undefined = this.trialSSHClientMap.get(trial.id);
+        if(!conn) {
+            throw new Error(`ssh client not set in trial ${trial.id}`);
+        }
+        sshClientManager.minusUsedConnectionNumber(conn);
```
Comment (yds05) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: user connection number is your implementation detail of sshClientManager, the caller doesn't need to know this. I suggest to rename minusUsedConnectionNumber to releaseConnection.

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 8]
```diff
@@ -115,6 +118,50 @@ class RemoteMachineTrainingService implements TrainingService {
...
+        this.machineSSHClientMap.set(trial.rmMeta, sshClientManager);
```
Comment (yds05) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: Useless line.

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 9]
```diff
@@ -430,9 +482,10 @@ class RemoteMachineTrainingService implements TrainingService {
             } else {
                 deferred.reject(new Error(`No valid passwd or sshKeyPath is configed.`));
             }
-            this.machineSSHClientMap.set(rmMeta, conn);
+            
             conn.on('ready', async () => {
-                this.machineSSHClientMap.set(rmMeta, conn);
+                //Initialize connection, the first connection is used for gpu collector
```
Comment (yds05) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: Consider to move all the ssh connection related code into SSHClientManager

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts (hunk 5 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: changed to `allocateSSHClientForTrial`

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts (hunk 6 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: Good suggestion, I've put `iniNewSSHClient()` into `getAvailableSSHClient()`

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts (hunk 7 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: changed to `releaseConnection`

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts (hunk 8 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: removed.

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts (hunk 9 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: updated.

File: src/nni_manager/training_service/remote_machine/remoteMachineData.ts [hunk 10]
```diff
@@ -94,6 +97,138 @@ export class RemoteMachineTrialJobDetail implements TrialJobDetail {
...
+        } else if(this.rmMeta.sshKeyPath) {
+            if(!fs.existsSync(this.rmMeta.sshKeyPath)) {
+                //SSh key path is not a valid file, reject
+                deferred.reject(new Error(`${this.rmMeta.sshKeyPath} does not exist.`));
+            }
+            const privateKey: string = fs.readFileSync(this.rmMeta.sshKeyPath, 'utf8');
+
+            connectConfig.privateKey = privateKey;
+            connectConfig.passphrase = this.rmMeta.passphrase;
+        } else {
```
Comment (chicm-ms) [inline src/nni_manager/training_service/remote_machine/remoteMachineData.ts]: no await for async function

Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/remoteMachineData.ts]: removed unused async

**Pull request merged**